File: server/app/routes/ws_agent.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from server.app.config import settings
from server.app.protocol import envelope, receive_json, send_json
from server.app.runtime import sock
from server.app.state import state
from server.app.services.agents import agent_connections
from server.app.services.events import update_node
from server.app.services.turns import mark_turn_completed, mark_turn_started


def running_turn_count(payload: dict[str, Any]) -> int:
    running_turn_ids = payload.get("running_turn_ids")
    if isinstance(running_turn_ids, list):
        return len(running_turn_ids)

    legacy_running_turns = payload.get("running_turns")
    if isinstance(legacy_running_turns, list):
        return len(legacy_running_turns)
    if isinstance(legacy_running_turns, int):
        return legacy_running_turns
    return 0


def connection_status(node_id: str) -> str:
    node = state.get_node(node_id) or {}
    if not node.get("approved", False):
        return "pending"
    return "online"


@sock.route("/ws/agent")
def agent_socket(ws: Any) -> None:
    node_id = "unbound"

    try:
        hello_message = receive_json(ws)
        if hello_message.get("type") != "agent.hello":
            send_json(
                ws,
                envelope(
                    "server.error",
                    {"code": "INVALID_HANDSHAKE", "message": "expected agent.hello"},
                    source_kind="server",
                    source_id="main",
                    target_kind="agent",
                    target_id=node_id,
                ),
            )
            return

        send_json(
            ws,
            envelope(
                "server.hello",
                {
                    "server_version": settings.app_version,
                    "heartbeat_sec": 15,
                    "resume_supported": False,
                },
                source_kind="server",
                source_id="main",
                target_kind="agent",
                target_id=node_id,
                request_id=hello_message.get("request_id"),
            ),
        )

        auth_message = receive_json(ws)
        token = auth_message.get("data", {}).get("pair_token")
        if auth_message.get("type") != "agent.auth" or token != settings.default_pair_token:
            send_json(
                ws,
                envelope(
                    "server.auth.reject",
                    {"code": "PAIR_TOKEN_INVALID", "message": "配对令牌无效或已过期"},
                    source_kind="server",
                    source_id="main",
                    target_kind="agent",
                    target_id=node_id,
                    request_id=auth_message.get("request_id"),
                ),
            )
            return

        hello_payload = hello_message.get("data", {})
        hostname = str(hello_payload.get("hostname", "local-node"))
        username = str(hello_payload.get("username", hostname))
        node_id = auth_message.get("data", {}).get("node_id") or f"node_{hostname}"
        existing_node = state.get_node(node_id) or {}
        agent_connections.upsert(node_id, ws)
        update_node(
            node_id,
            {
                "node_id": node_id,
                "name": username,
                "hostname": hostname,
                "platform": hello_payload.get("platform"),
                "arch": hello_payload.get("arch"),
                "agent_version": hello_payload.get("agent_version"),
                "hello_message": hello_payload.get("hello_message") or existing_node.get("hello_message", ""),
                "approved": bool(existing_node.get("approved", False)),
                "status": "online" if existing_node.get("approved", False) else "pending",
                "running_turns": 0,
                "worker_count": 0,
            },
        )

        send_json(
            ws,
            envelope(
                "server.auth.ok",
                {
                    "node_id": node_id,
                    "node_name": username,
                    "max_workers": 2,
                },
                source_kind="server",
                source_id="main",
                target_kind="agent",
                target_id=node_id,
                request_id=auth_message.get("request_id"),
            ),
        )

        while True:
            message = receive_json(ws)
            message_type = message.get("type")
            payload = message.get("data", {})

            if message_type == "agent.state.report":
                update_node(
                    node_id,
                    {
                        "status": connection_status(node_id),
                        "running_turns": running_turn_count(payload),
                        "worker_count": payload.get("worker_count", running_turn_count(payload)),
                    },
                )
                continue

            if message_type == "agent.ping":
                update_node(
                    node_id,
                    {
                        "status": connection_status(node_id),
                        "running_turns": running_turn_count(payload),
                        "worker_count": payload.get("worker_count", 0),
                    },
                )
                send_json(
                    ws,
                    envelope(
                        "server.pong",
                        {"server_time": datetime.now(tz=UTC).isoformat()},
                        source_kind="server",
                        source_id="main",
                        target_kind="agent",
                        target_id=node_id,
                        request_id=message.get("request_id"),
                    ),
                )
                continue

            if message_type == "agent.turn.started":
                turn_id = str(payload.get("turn_id", ""))
                if turn_id:
                    mark_turn_started(turn_id=turn_id, node_id=node_id)
                update_node(
                    node_id,
                    {
                        "status": connection_status(node_id),
                        "running_turns": running_turn_count(payload),
                        "worker_count": payload.get("worker_count", running_turn_count(payload)),
                    },
                )
                continue

            if message_type == "agent.turn.completed":
                turn_id = str(payload.get("turn_id", ""))
                output = str(payload.get("output", "")).strip()
                if turn_id and output:
                    mark_turn_completed(turn_id=turn_id, node_id=node_id, output=output)
                update_node(
                    node_id,
                    {
                        "status": connection_status(node_id),
                        "running_turns": running_turn_count(payload),
                        "worker_count": payload.get("worker_count", 0),
                    },
                )
                continue

            send_json(
                ws,
                envelope(
                    "server.notice",
                    {"level": "info", "message": f"server ignored message type: {message_type}"},
                    source_kind="server",
                    source_id="main",
                    target_kind="agent",
                    target_id=node_id,
                    request_id=message.get("request_id"),
                ),
            )
    except Exception:
        agent_connections.remove(node_id)
        disconnected_status = "offline" if (state.get_node(node_id) or {}).get("approved", False) else "pending"
        update_node(node_id, {"status": disconnected_status})

File: cmd/qunce/ws_agent.go
package main

import (
	"log"
	"net/http"
	"strings"

	"github.com/gorilla/websocket"
)

var wsAgentUpgrader = websocket.Upgrader{
	CheckOrigin: func(_ *http.Request) bool {
		return true
	},
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
}

func registerAgentSocket(mux *http.ServeMux, st *state, agents *agentRegistry, consoles *consoleRegistry, cfg appConfig) {
	mux.HandleFunc("/ws/agent", handleAgentSocket(st, agents, consoles, cfg))
}

func handleAgentSocket(st *state, agents *agentRegistry, consoles *consoleRegistry, cfg appConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		conn, err := wsAgentUpgrader.Upgrade(w, r, nil)
		if err != nil {
			log.Printf("upgrade agent socket: %v", err)
			return
		}

		nodeID := "unbound"
		defer func() {
			agents.remove(nodeID)
			node := st.getNode(nodeID)
			if node != nil {
				status := "pending"
				if node.Approved {
					status = "offline"
				}
				st.updateNode(nodeID, map[string]interface{}{
					"status": status,
				})
				broadcastNodeUpdate(consoles, st)
			}
			_ = conn.Close()
		}()

		helloEnvelope, err := readEnvelope(conn)
		if err != nil {
			_ = conn.WriteMessage(websocket.TextMessage, mustJSON(buildEnvelope(
				"server.error",
				"server",
				"main",
				"agent",
				nodeID,
				map[string]interface{}{
					"code":    "INVALID_HANDSHAKE",
					"message": "expected agent.hello",
				},
				"",
			)))
			return
		}
		if helloEnvelope.Type != "agent.hello" {
			_ = conn.WriteMessage(websocket.TextMessage, mustJSON(buildEnvelope(
				"server.error",
				"server",
				"main",
				"agent",
				nodeID,
				map[string]interface{}{
					"code":    "INVALID_HANDSHAKE",
					"message": "expected agent.hello",
				},
				helloEnvelope.RequestID,
			)))
			return
		}

		helloPayload, err := dataMap(helloEnvelope.Data)
		if err != nil {
			_ = conn.WriteMessage(websocket.TextMessage, mustJSON(buildEnvelope(
				"server.error",
				"server",
				"main",
				"agent",
				nodeID,
				map[string]interface{}{
					"code":    "INVALID_HANDSHAKE",
					"message": "invalid hello payload",
				},
				helloEnvelope.RequestID,
			)))
			return
		}

		if err := sendEnvelope(conn, buildEnvelope(
			"server.hello",
			"server",
			"main",
			"agent",
			nodeID,
			map[string]interface{}{
				"server_version":   cfg.AppVersion,
				"heartbeat_sec":    15,
				"resume_supported": false,
			},
			helloEnvelope.RequestID,
		)); err != nil {
			return
		}

		authEnvelope, err := readEnvelope(conn)
		if err != nil {
			_ = conn.WriteMessage(websocket.TextMessage, mustJSON(buildEnvelope(
				"server.auth.reject",
				"server",
				"main",
				"agent",
				nodeID,
				map[string]interface{}{
					"code":    "INVALID_AUTH",
					"message": "missing auth message",
				},
				"",
			)))
			return
		}
		authPayload, err := dataMap(authEnvelope.Data)
		if err != nil {
			_ = conn.WriteMessage(websocket.TextMessage, mustJSON(buildEnvelope(
				"server.auth.reject",
				"server",
				"main",
				"agent",
				nodeID,
				map[string]interface{}{
					"code":    "INVALID_AUTH",
					"message": "invalid auth payload",
				},
				authEnvelope.RequestID,
			)))
			return
		}

		if authEnvelope.Type != "agent.auth" || toString(authPayload["pair_token"], "") != cfg.DefaultPairToken {
			_ = conn.WriteMessage(websocket.TextMessage, mustJSON(buildEnvelope(
				"server.auth.reject",
				"server",
				"main",
				"agent",
				nodeID,
				map[string]interface{}{
					"code":    "PAIR_TOKEN_INVALID",
					"message": "pair token invalid",
				},
				authEnvelope.RequestID,
			)))
			return
		}

		hostname := strings.TrimSpace(toString(helloPayload["hostname"], "local-node"))
		username := strings.TrimSpace(toString(helloPayload["username"], hostname))
		nodeID = strings.TrimSpace(toString(authPayload["node_id"], ""))
		if nodeID == "" {
			nodeID = "node_" + strings.ReplaceAll(hostname, " ", "-")
		}
		existingNode := st.getNode(nodeID)
		if existingNode == nil {
			existingNode = &nodeRecord{}
		}

		agents.upsert(nodeID, conn)
		st.updateNode(nodeID, map[string]interface{}{
			"node_id":       nodeID,
			"name":          username,
			"hostname":      hostname,
			"work_dir":      firstNonEmpty(toString(helloPayload["work_dir"], ""), existingNode.WorkDir),
			"platform":      firstNonEmpty(toString(helloPayload["platform"], ""), existingNode.Platform),
			"arch":          firstNonEmpty(toString(helloPayload["arch"], ""), existingNode.Arch),
			"agent_version": firstNonEmpty(toString(helloPayload["agent_version"], ""), existingNode.AgentVersion),
			"hello_message": firstNonEmpty(toString(helloPayload["hello_message"], ""), existingNode.HelloMessage),
			"approved":      existingNode.Approved,
			"status":        boolText(existingNode != nil && existingNode.Approved, "online", "pending"),
			"running_turns": 0,
			"worker_count":  0,
		})
		broadcastNodeUpdate(consoles, st)

		if err := sendEnvelope(conn, buildEnvelope(
			"server.auth.ok",
			"server",
			"main",
			"agent",
			nodeID,
			map[string]interface{}{
				"node_id":     nodeID,
				"node_name":   username,
				"max_workers": 2,
			},
			authEnvelope.RequestID,
		)); err != nil {
			return
		}

		_ = dispatchPendingTurnsForNode(st, agents, nodeID)

		sendEnvelopeSafeForNode := func(target *agentConnection, data envelope) error {
			if target == nil {
				return nil
			}
			return sendEnvelopeSafe(target, data)
		}

		for {
			raw, err := readEnvelope(conn)
			if err != nil {
				return
			}
			payload, err := dataMap(raw.Data)
			if err != nil {
				_ = sendEnvelopeSafeForNode(agents.get(nodeID), buildEnvelope(
					"server.notice",
					"server",
					"main",
					"agent",
					nodeID,
					map[string]interface{}{"level": "warning", "message": "invalid payload"},
					raw.RequestID,
				))
				continue
			}

			updateNodeState := func() {
				runningTurns := toInt(payload["running_turns"])
				if runningTurns == 0 {
					if ids, ok := payload["running_turn_ids"].([]interface{}); ok {
						runningTurns = len(ids)
					}
				}
				_ = st.updateNode(nodeID, map[string]interface{}{
					"status":        connectionStatus(st.getNode(nodeID)),
					"running_turns": runningTurns,
					"worker_count":  toInt(payload["worker_count"]),
				})
			}

			switch raw.Type {
			case "agent.state.report":
				updateNodeState()
			case "agent.ping":
				updateNodeState()
				_ = sendEnvelopeSafeForNode(agents.get(nodeID), buildEnvelope(
					"server.pong",
					"server",
					"main",
					"agent",
					nodeID,
					map[string]interface{}{"server_time": nowRFC3339()},
					raw.RequestID,
				))
			case "agent.turn.started":
				turnID := strings.TrimSpace(toString(payload["turn_id"], ""))
				if turnID != "" {
					markTurnStarted(st, turnID, nodeID)
				}
				updateNodeState()
			case "agent.turn.read":
				turnID := strings.TrimSpace(toString(payload["turn_id"], ""))
				if turnID != "" {
					markTurnRead(st, turnID, nodeID)
					if turn := st.getTurn(turnID); turn != nil {
						broadcastChatSnapshot(consoles, st, turn.ChatID)
					}
				}
				updateNodeState()
			case "agent.turn.completed":
				turnID := strings.TrimSpace(toString(payload["turn_id"], ""))
				output := strings.TrimSpace(toString(payload["output"], ""))
				if turnID != "" && output != "" {
					markTurnCompleted(st, turnID, nodeID, output)
				}
				updateNodeState()
			case "agent.workspace.validated":
				requestID := strings.TrimSpace(raw.RequestID)
				if requestID != "" {
					agents.resolveWorkspaceValidation(requestID, map[string]interface{}{
						"ok":              toBool(payload["ok"], false),
						"normalized_path": strings.TrimSpace(toString(payload["normalized_path"], "")),
						"message":         strings.TrimSpace(toString(payload["message"], "")),
					})
				}
			default:
				_ = sendEnvelopeSafeForNode(agents.get(nodeID), buildEnvelope(
					"server.notice",
					"server",
					"main",
					"agent",
					nodeID,
					map[string]interface{}{"level": "info", "message": "server ignored message type: " + raw.Type},
					raw.RequestID,
				))
			}
		}
	}
}

func connectionStatus(node *nodeRecord) string {
	if node == nil {
		return "offline"
	}
	if node.Approved {
		return "online"
	}
	return "pending"
}
